### PR TITLE
[primTorch] move diagonal & add linalg.diagonal refs

### LINKED ISSUE #85419: primTorch: move `linalg.diagonal` out of `diagonal` `OpInfo`
### 🐛 Describe the bug

```py
    OpInfo('diagonal',
           # TODO: They are not strictly aliases as they have diverging
           # defaults, but we can see them as aliases for testing purposes.  If
           # we add tests that test the function against the alias, make
           # linalg.diagonal into its own OpInfo
           aliases=('linalg.diagonal',),
```

### Versions

master

cc @ezyang @mruberry @ngimel @Lezcano @fdrocha

### PR DESCRIPTION
**Pull request by khushi-411**
Fixes #85419

Also, add `_refs.linalg.diagonal`.

cc @lezcano @kshitij12345

### CODE REVIEW

File: torch/_refs/linalg/__init__.py [hunk 1]
```diff
@@ -57,6 +58,16 @@ def check_norm_dtype(dtype: Optional[torch.dtype], x_dtype: torch.dtype, fn_name
 from torch._decomp import register_decomposition
 
 
+def diagonal(
+    input: TensorLikeType,
+    *,
+    offset: int = 0,
+    dim1: int = -2,
```
Comment (khushi-411) [inline torch/_refs/linalg/__init__.py:65]: cc'ing @nkaretnikov, who originally added this ref in #84792. I added this while working on #85419. Hope it is fine. :) 
 Thanks!

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 2]
```diff
@@ -1172,6 +1249,26 @@ def make_input():
...
+        supports_forward_ad=True,
+        supports_fwgrad_bwgrad=True,
+        sample_inputs_func=sample_inputs_diagonal,
+        error_inputs_func=error_inputs_diagonal,
+        skips=(
+            DecorateInfo(
+                unittest.skip("Skipped"),
+                "TestVmapOperatorsOpInfo",
+                "test_vmap_linalg_failure_1D_input",
+            ),
```
Comment (khushi-411) [inline torch/testing/_internal/opinfo/definitions/linalg.py:1269]: This seems like a bug on the functorch side:
```python
In [1]: import torch

In [2]: from torch import vmap

In [3]: a = torch.randn(3, 3)

In [4]: torch.linalg.diagonal(a)
Out[4]: tensor([ 0.4697, -1.6019, -0.6110])

In [5]: vmap(torch.linalg.diagonal(a))
Out[5]: <function torch.vmap.<locals>.wrapped>

In [6]: vmap(torch.linalg.diagonal)(a)
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 vmap(torch.linalg.diagonal)(a)

File ~/Documents/ldev/pytorch/torch/_functorch/vmap.py:434, in vmap.<locals>.wrapped(*args, **kwargs)
    430     return _chunked_vmap(func, flat_in_dims, chunks_flat_args,
    431                          args_spec, out_dims, randomness, **kwargs)
    433 # If chunk_size is not specified.
--> 434 return _flat_vmap(
    435     func, batch_size, flat_in_dims, flat_args, args_spec, out_dims, randomness, **kwargs
    436 )

File ~/Documents/ldev/pytorch/torch/_functorch/vmap.py:39, in doesnt_support_saved_tensors_hooks.<locals>.fn(*args, **kwargs)
     36 @functools.wraps(f)
     37 def fn(*args, **kwargs):
     38     with torch.autograd.graph.disable_saved_tensors_hooks(message):
---> 39         return f(*args, **kwargs)

File ~/Documents/ldev/pytorch/torch/_functorch/vmap.py:619, in _flat_vmap(func, batch_size, flat_in_dims, flat_args, args_spec, out_dims, randomness, **kwargs)
    617 try:
    618     batched_inputs = _create_batched_inputs(flat_in_dims, flat_args, vmap_level, args_spec)
--> 619     batched_outputs = func(*batched_inputs, **kwargs)
    620     return _unwrap_batched(batched_outputs, out_dims, vmap_level, batch_size, func)
    621 finally:

IndexError: Dimension out of range (expected to be in range of [-1, 0], but got -2)
```

Comment (lezcano) [reply torch/testing/_internal/opinfo/definitions/linalg.py:1269]: If it's an error (which it looks like it is) xfail it. Also, the list of xfails for functorch ops is located in the test itself. In this case, it's in https://github.com/pytorch/pytorch/blob/e3892fd16be708ade8e6bbb31e9c0b3771d24fbb/test/functorch/test_vmap.py#L4251

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 3]
```diff
@@ -1172,6 +1249,26 @@ def make_input():
             ),
         ),
     ),
+    OpInfo(
+        "linalg.diagonal",
+        aten_name="lingalg_diagonal",
```
Comment (lezcano) [inline torch/testing/_internal/opinfo/definitions/linalg.py:1254]: errata

Comment (khushi-411) [reply torch/testing/_internal/opinfo/definitions/linalg.py:1254]: Whoops. Thanks!

File: torch/testing/_internal/opinfo/definitions/linalg.py (hunk 2 above)
Comment (khushi-411) [reply torch/testing/_internal/opinfo/definitions/linalg.py:1269]: Sure, updated!

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 4]
```diff
@@ -790,6 +790,83 @@ def error_inputs_lstsq_grad_oriented(op_info, device, **kwargs):
...
+            err_dim = dim1 if dim1_cond else dim2
+            err = (
+                r"Dimension out of range \(expected to be in range of "
+                rf"\[{bound1}, {bound2}\], but got {err_dim}\)"
+            )
+            yield ErrorInput(sample, error_regex=err, error_type=IndexError)
+        else:
+            raise RuntimeError("should be unreachable")
+
+
```
Comment (lezcano) [inline torch/testing/_internal/opinfo/definitions/linalg.py:869]: Any reason not to simply import these from `common_method_invocations`?

File: torch/testing/_internal/opinfo/definitions/linalg.py [hunk 5]
```diff
@@ -790,6 +790,24 @@ def error_inputs_lstsq_grad_oriented(op_info, device, **kwargs):
...
+def error_inputs_diagonal(op_info, device, **kwargs):
+    from torch.testing._internal.common_methods_invocations import (
+        error_inputs_diagonal_diag_embed,
+    )
+
+    return error_inputs_diagonal_diag_embed(op_info, device, **kwargs)
+
+
```
[the remark this answers is not in the document]
Comment (khushi-411) [reply torch/testing/_internal/opinfo/definitions/linalg.py:810]: Yeah, I agree. But when I imported on top, it caused a circular import error.

[the remark this answers is not in the document]
Comment (lezcano) [reply torch/testing/_internal/opinfo/definitions/linalg.py:810]: Right. That means that the definitions should live in `linalg.py` (with its name changed of course), and then this should be imported from `common_methods_invocations.py`

**Comment by khushi-411** [comment]
Hi,
@lezcano, gentle ping. Would you like for me to add anything else? If it's okay, please approve. Thanks!

**Comment by lezcano** [comment]
This comment https://github.com/pytorch/pytorch/pull/95774#discussion_r1121872855 was not addressed. The proper way of doing this is to have the samples in the linalg file and import them from the ops file, as we do with other linalg functions like svd, cholesky, etc

**Comment by khushi-411** [comment]
Ah, I'm sorry. I've updated it now. Thanks for reclarifying 😃

**Review by lezcano — APPROVED** [review]
Thank you!

**Pull request closed without merging**